assets/tests/TaskSchemaValidationServiceSpec.js

```diff
@@ -0,0 +1,301 @@
+// Copyright 2017 The TIE Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS-IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * @fileoverview Unit tests for the QuestionSchemaValidatorService.
+ */
+
+describe('TaskSchemaValidationService', function() {
+  var QuestionDataService;
+  // TaskSchemaValidationService needs a shorter name.
+  var Tsvs;
+  var questions = [];
+  // Should contain all question IDs.
+  // TODO(eyurko): Figure out a way to dynamically check to make sure
+  // that all question IDs are specified.
+  var questionIds = ['reverseWords', 'parens', 'i18n', 'rle'];
+  globalData.questionSets.all = {};
+  globalData.questionSets.all.questionIds = questionIds;
+  globalData.questionSets.all.introductionParagraphs = [];
+  beforeEach(module('tie'));
+  beforeEach(inject(function($injector) {
+    // Used for testing the validator. Values will be inserted during the tests
+    // so that we don't have to redefine the dict every time.
+    QuestionDataService = $injector.get('QuestionDataService');
+    Tsvs = $injector.get('TaskSchemaValidationService');
+    QuestionDataService.initCurrentQuestionSet('all');
+    questions = [];
+    questionIds.forEach(function(item) {
+      questions.push(QuestionDataService.getQuestion(item));
+    });
+  }));
+
+  var assertTrueOnAllTasks = function(func, message) {
+    questions.forEach(function(item) {
+      var tasks = item.getTasks();
+      for (var i = 0; i < tasks.length; i++) {
+        if (message) {
+          expect(func(tasks[i])).toBe(true,
+            [item.getTitle(), message, i].join(''));
```

Comment:
You could, e.g., have something in this file that's like (nb: this is a sketch, and is untested):

```
  describe('validateAllTasks', function() {
    it('should validate the structure of all sample tasks', function() {
        questions.forEach(function(question) {
          question.getTasks().forEach(function(task) {
            for (var verifierFunc in Tsvs) {
              verifierFunc(task, starterCode, auxiliaryCode);
            }
         });
      });
  });
```

A side-effect of this is that all your verifiers need to have the same function signature, even those that don't use starterCode or auxiliaryCode. We could leave those undefined in some cases, but this feels a bit icky. So another approach might be to implement an init(), viz.:

```
  describe('validateAllTasks', function() {
    it('should validate the structure of all sample tasks', function() {
        Tsvs.init(starterCode, auxiliaryCode);
        questions.forEach(function(question) {
          question.getTasks().forEach(function(task) {
            for (var verifierFunc in Tsvs) {
              verifierFunc(task);
            }
         });
      });
  });
```

Of course, you'd need to test this to make sure it outputs errors correctly, but if you can make it work then I think it should save a lot of maintenance.

Does this help?